rfc/20250211-s3-locking-with-conditional-writes.md

```diff
@@ -0,0 +1,133 @@
+> 
+> The `If-None-Match` header prevents overwrites of existing data by validating that there's not an object with the same key name already in your bucket.
+>
+> Alternatively, you can add the `If-Match` header to check an object's entity tag (`ETag`) before writing an object. With this header, Amazon S3 compares the provided `ETag` value with the `ETag` value of the object in S3. If the `ETag` values don't match, the operation fails.
+
+
+To allow this in OpenTofu, the `backend "s3"` will receive one new attribute:
+* `use_lockfile` `bool` (Default: `false`) - Flag to indicate if the locking should be performed by using strictly the S3 bucket.
+
+> [!NOTE]
+>
+> The name `use_lockfile` was selected this way to keep the [feature parity with Terraform](https://developer.hashicorp.com/terraform/language/backend/s3#state-locking).
+
+### User Documentation
+
+In order to make use of this new feature, the users will have to add the attribute in the `backend` block:
+```terraform
+terraform {
+  backend "s3" {
+    bucket = "tofu-state-backend"
+    key = "statefile"
+    region = "us-east-1"
+    dynamodb_table = "tofu_locking"
+    use_lockfile = true
+  }
+}
+```
+
+* When the new attribute `use_lockfile` exists and `dynamodb_table` is missing, OpenTofu will try to acquire the lock inside the configured S3 bucket.
+* When the new attribute `use_lockfile` will exist **alongside** `dynamodb_table`, OpenTofu will:
+  * Acquire the lock in the S3 bucket;
+  * Acquire the lock in DynamoDB table;
+  * Get the digest of the state object from DynamoDB and ensure the state object content integrity;
+  * Perform the requested sub-command;
+  * Release the lock from the S3 bucket;
+  * Release the lock from DynamoDB table.
+
+The usage of [workspaces](https://opentofu.org/docs/language/state/workspaces/) will not impact this new way of locking. The locking object will always be stored right next to its related state object.
+
+> [!NOTE]
+>
+> OpenTofu [recommends](https://opentofu.org/docs/language/settings/backends/s3/) to have versioning enabled for the S3 buckets used to store state objects.
+>
+> Acquiring and releasing locks will add a good amount of writes and reads to the bucket. Therefore, for a versioning-enabled S3 bucket, the number of versions for that object could grow significantly.
+> Even though the cost should be negligible for the locking objects, any user using this feature could consider configuring the lifecycle of the S3 bucket to limit the number of versions of an object.
+
+> [!WARNING]
+> 
+> When OpenTofu S3 backend is used with an S3 compatible provider, it needs to be checked that the provider is supporting conditional writes in the same way AWS S3 is offering.
+
+#### Migration paths
+##### I have no locking enabled
+In this case, the user can just add the new `use_lockfile=true` and run `tofu init -reconfigure`.
+
+##### I have DynamoDB locking enabled
+In case the user is having DynamoDB enabled, there are two paths forward:
```

Comment:
```suggestion
In case the user has DynamoDB enabled, there are two paths forward:
```

Reply:
Applied in 6c43f94d46a40ebf22b652cb08d42bd5ba7f4179